File: src/main/java/net/minecraft/client/particle/FlameParticle.java
package net.minecraft.client.particle;

import net.minecraft.client.renderer.Tesselator;
import net.minecraft.world.level.Level;

public class FlameParticle extends Particle {
   private float oSize;

   public FlameParticle(Level level, double x, double y, double z, double xd, double yd, double zd) {
      super(level, x, y, z, xd, yd, zd);
      this.xd = this.xd * (double)0.01F + xd;
      this.yd = this.yd * (double)0.01F + yd;
      this.zd = this.zd * (double)0.01F + zd;
      double var10000 = x + (double)((this.random.nextFloat() - this.random.nextFloat()) * 0.05F);
      var10000 = y + (double)((this.random.nextFloat() - this.random.nextFloat()) * 0.05F);
      var10000 = z + (double)((this.random.nextFloat() - this.random.nextFloat()) * 0.05F);
      this.oSize = this.size;
      this.rCol = this.gCol = this.bCol = 1.0F;
      this.lifetime = (int)((double)8.0F / (Math.random() * 0.8 + 0.2)) + 4;
      this.noPhysics = true;
      this.tex = 48;
   }

   public void render(Tesselator t, float a, float xa, float ya, float za, float xa2, float za2) {
      float s = ((float)this.age + a) / (float)this.lifetime;
      this.size = this.oSize * (1.0F - s * s * 0.5F);
      super.render(t, a, xa, ya, za, xa2, za2);
   }

   public float getBrightness(float a) {
      float l = ((float)this.age + a) / (float)this.lifetime;
      if (l < 0.0F) {
         l = 0.0F;
      }

      if (l > 1.0F) {
         l = 1.0F;
      }

      float br = super.getBrightness(a);
      return br * l + (1.0F - l);
   }

   public void tick() {
      this.xo = this.x;
      this.yo = this.y;
      this.zo = this.z;
      if (this.age++ >= this.lifetime) {
         this.remove();
      }

      this.move(this.xd, this.yd, this.zd);
      this.xd *= (double)0.96F;
      this.yd *= (double)0.96F;
      this.zd *= (double)0.96F;
      if (this.onGround) {
         this.xd *= (double)0.7F;
         this.zd *= (double)0.7F;
      }

   }
}

File: mc-cpp/src/particle/FlameParticle.cpp
#include "particle/FlameParticle.hpp"
#include "world/Level.hpp"

namespace mc {

FlameParticle::FlameParticle(Level* level, double x, double y, double z, double xa, double ya, double za)
    : Particle(level, x, y, z, 0.0, 0.0, 0.0)
{
    auto randFloat = [this]() { return static_cast<float>(random()) / random.max(); };

    // Override velocity with smaller values (matching Java FlameParticle)
    xd = xd * 0.01f + xa;
    yd = yd * 0.01f + ya;
    zd = zd * 0.01f + za;

    // Small random position offset (Java does this but doesn't use the result)
    // x + (randFloat() - randFloat()) * 0.05f;
    // y + (randFloat() - randFloat()) * 0.05f;
    // z + (randFloat() - randFloat()) * 0.05f;

    // Store original size
    oSize = size;

    // White/yellow color (matching Java: rCol = gCol = bCol = 1.0F)
    rCol = gCol = bCol = 1.0f;

    // Lifetime: 8-12 ticks (matching Java: 8.0 / (random * 0.8 + 0.2) + 4)
    lifetime = static_cast<int>(8.0 / (randFloat() * 0.8 + 0.2)) + 4;

    // No physics (matching Java: noPhysics = true)
    // We'll handle this by not calling move() with collision

    // Texture index 48 (flame texture in particles.png)
    tex = 48;
}

void FlameParticle::tick() {
    prevX = x;
    prevY = y;
    prevZ = z;

    if (age++ >= lifetime) {
        remove();
        return;
    }

    // Move without collision (noPhysics = true in Java)
    x += xd;
    y += yd;
    z += zd;

    // Apply drag (matching Java: 0.96F)
    xd *= 0.96;
    yd *= 0.96;
    zd *= 0.96;

    // Ground friction (shouldn't happen often since noPhysics, but included for completeness)
    if (onGround) {
        xd *= 0.7;
        zd *= 0.7;
    }
}

float FlameParticle::getRenderSize(float partialTick) const {
    // Matching Java FlameParticle.render():
    // float s = ((float)this.age + a) / (float)this.lifetime;
    // this.size = this.oSize * (1.0F - s * s * 0.5F);
    float s = (static_cast<float>(age) + partialTick) / static_cast<float>(lifetime);
    return oSize * (1.0f - s * s * 0.5f);
}

} // namespace mc
